tests/test_filters_first.py

```diff
@@ -0,0 +1,243 @@
+
+        rusty_engine = engines["rusty"]
+        django_engine = engines["django"]
+
+        # Match Django's test exactly
+        template = "{{ a|first }} {{ b|first }}"
+        context = {"a": ["a&b", "x"], "b": [mark_safe("a&b"), "x"]}
+
+        rusty_result = rusty_engine.from_string(template).render(context)
+        django_result = django_engine.from_string(template).render(context)
+
+        assert rusty_result == "a&amp;b a&b"
+        assert django_result == "a&amp;b a&b"
+
+    def test_first_autoescape_off(self):
+        """Test first filter with autoescape off."""
+        rusty_engine = engines["rusty"]
+        django_engine = engines["django"]
+
+        template = "{% autoescape off %}{{ items|first }}{% endautoescape %}"
+        context = {"items": ["<b>bold</b>", "text"]}
+
+        rusty_result = rusty_engine.from_string(template).render(context)
+        django_result = django_engine.from_string(template).render(context)
+
+        assert rusty_result == "<b>bold</b>"
+        assert django_result == "<b>bold</b>"
+
+    def test_first_autoescape_off_with_safe(self):
+        """Test with autoescape off for both escaped and safe strings (Django's first02)."""
+        from django.utils.safestring import mark_safe
+
+        rusty_engine = engines["rusty"]
+        django_engine = engines["django"]
+
+        # Match Django's test exactly
+        template = "{% autoescape off %}{{ a|first }} {{ b|first }}{% endautoescape %}"
+        context = {"a": ["a&b", "x"], "b": [mark_safe("a&b"), "x"]}
+
+        rusty_result = rusty_engine.from_string(template).render(context)
+        django_result = django_engine.from_string(template).render(context)
+
+        assert rusty_result == "a&b a&b"
+        assert django_result == "a&b a&b"
+
+    def test_first_with_safe_string(self):
+        """Test first filter with mark_safe string."""
+        from django.utils.safestring import mark_safe
+
+        rusty_engine = engines["rusty"]
+        django_engine = engines["django"]
+
+        template = "{{ items|first }}"
+        context = {"items": [mark_safe("<b>bold</b>"), "text"]}
+
+        # When the first item is marked safe, it should not be escaped
+        rusty_result = rusty_engine.from_string(template).render(context)
+        django_result = django_engine.from_string(template).render(context)
+
+        assert rusty_result == "<b>bold</b>"
+        assert django_result == "<b>bold</b>"
+
+    def test_first_no_argument(self):
+        """Test that first filter doesn't accept arguments."""
+        rusty_engine = engines["rusty"]
+        django_engine = engines["django"]
+
+        template = "{{ items|first:'arg' }}"
+        context = {"items": [1, 2, 3]}
+
+        with pytest.raises(TemplateSyntaxError):
+            rusty_engine.from_string(template).render(context)
+        with pytest.raises(TemplateSyntaxError):
+            django_engine.from_string(template).render(context)
```

Comment:
When testing `TemplateSyntaxError` I like to also check the error message provided by Django and by our code. You can find examples in the other test files.

If you add the custom `RenderError` variants for `TypeError`, I'd also like to see the same in those tests.